File: advent-of-code-2023/day01/rust/src/main.rs
use std::fs::File;
use std::io::{self, BufRead};

const NUMS: [&str; 20] = [
    "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
    "0", "1", "2", "3", "4", "5", "6", "7", "8", "9",
];

fn main() {
    let fname = "puzzleInput.txt";

    let file = File::open(fname).expect("Could not open file");
    let lines = io::BufReader::new(file).lines();

    // let total_part1 = lines.fold(0u32, |a, x| a + part1(x.unwrap()));
    // println!("{}", total_part1);

    let total_part2 = lines.fold(0u32, |a, x| a + part2(x.unwrap()));
    println!("{}", total_part2);
}

fn part1(line: String) -> u32 {
    let first = line
        .chars()
        .find(|&x| x.is_numeric())
        .expect("No digit found");
    let last = line
        .chars()
        .rev()
        .find(|&x| x.is_numeric())
        .expect("No digit found");

    let first_last = format!("{}{}", first, last);

    return first_last.parse::<u32>().unwrap();
}

fn part2(line: String) -> u32 {
    let mut numbers: Vec<_> = vec![];

    for num in NUMS {
        numbers.append(&mut line.match_indices(num).into_iter().collect::<Vec<_>>())
    }

    let mut min_idx = usize::MAX;
    let mut max_idx = usize::MIN;
    let mut first: Option<&str> = None;
    let mut last: Option<&str> = None;
    for (idx, num) in numbers {
        if idx <= min_idx {
            min_idx = idx;
            first = Some(num);
        }
        if idx >= max_idx {
            max_idx = idx;
            last = Some(num);
        }
    }

    let first_index = digit_from_str(first.expect("First digit missing"));
    let last_index = digit_from_str(last.expect("Last digit missing"));

    let first_last = format!("{}{}", first_index, last_index);
    return first_last.parse::<u32>().unwrap();
}

fn digit_from_str(digit_str: &str) -> usize {
    let mut index = NUMS.iter().position(|&x| x == digit_str).unwrap();
    if index > 9 {
        index -= 10
    }

    return index;
}

File: advent-of-code-2023/day01/ocaml/bin/main.ml
let file = "../puzzleInput.txt"

let is_digit c =
    match c with '0' .. '9' -> true | _ -> false

let find_number line =
        let first_digit = line |> String.to_seq |> List.of_seq |> List.find (is_digit) in
        let last_digit = line |> String.to_seq |> List.of_seq |> List.rev |> List.find (is_digit) in
        let number = Printf.sprintf "%c%c" first_digit last_digit in
        let number = int_of_string number in
        number

let rec part1 channel acc =
    match input_line channel with
    | line -> 
        let acc = acc + find_number(line) in
        part1 channel acc

    | exception End_of_file -> acc

let () =
    let ic = open_in file in
    let part1_sum = part1 ic 0 in
    Printf.printf "Part 1: %d\n" part1_sum;

(* let rec part2 channel acc = *)
(*     match input_line channel with *)
(*     | line ->  *)
(*         let line = replace_words replacements line in *)
(*         let acc = acc + find_number(line) in *)
(*         part2 channel acc *)
(**)
(*     | exception End_of_file -> acc *)

(* let replacements = [ *)
(*     "zero", "z0e"; *)
(*     "one", "o1e"; *)
(*     "two", "t2o"; *)
(*     "three", "t3e"; *)
(*     "four", "f4r"; *)
(*     "five", "f5e"; *)
(*     "six", "s6x"; *)
(*     "seven", "s7n"; *)
(*     "eight", "e8t"; *)
(*     "nine", "n9e"; *)
(* ] *)
(**)
(* let rec replace_words replacements line = *)
(*      match replacements with *)
(*         | [] -> line *)
(*         | x :: xs -> String.replace x[0 x.1; replace_words replacements line  *)
(**)
